### PR TITLE
Created Flexbox mixin and added it to ComplexWidget

### PR DESCRIPTION
**Pull request by chriswjones**
This adds the option to use flexbox layout. Let me know if you have feedback.

### CODE REVIEW

File: gwt-material/src/main/java/gwt/material/design/client/constants/Display.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+ * limitations under the License.
+ * #L%
+ */
+
+import com.google.gwt.dom.client.Style;
+
+/**
+ * @author chriswjones
+ */
+public enum Display {
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/constants/Display.java]: We shouldn't reimplement Display, use a helper class instead.

File: gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+
+/**
+ * Mixin for the {@link MaterialTooltip} component.
+ *
+ * @author gilberto-torrezan
+ * @see HasTooltip
+ * @see ComplexWidget
+ */
+public class FlexboxMixin<T extends Widget & HasFlexbox> extends AbstractMixin<T> implements HasFlexbox {
+
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java]: Remove extra new line

File: gwt-material/src/main/java/gwt/material/design/client/base/ComplexWidget.java [hunk 3]
```diff
@@ -93,90 +76,131 @@ public void insert(final Widget child, final int beforeIndex) {
     }
 
     private IdMixin<ComplexWidget> getIdMixin() {
-        if(idMixin == null) { idMixin = new IdMixin<>(this); }
+        if (idMixin == null) {
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/base/ComplexWidget.java]: Please undo these changes. We are leaving them collapsed.

**Comment by BenDol** [comment]
I'm wondering if rather than adding this to the ComplexWidget if we create a seperate widget for Flexbox layouts. Rather than bloating the ComplexWidget even more.

File: gwt-material/src/main/java/gwt/material/design/client/base/ComplexWidget.java [hunk 4]
```diff
@@ -20,37 +20,19 @@
...
-
 import com.google.gwt.dom.client.Element;
 import com.google.gwt.dom.client.Style;
 import com.google.gwt.user.client.ui.ComplexPanel;
 import com.google.gwt.user.client.ui.Focusable;
 import com.google.gwt.user.client.ui.HasEnabled;
 import com.google.gwt.user.client.ui.Widget;
+import gwt.material.design.client.base.helper.StyleHelper;
+import gwt.material.design.client.base.mixin.*;
+import gwt.material.design.client.constants.*;
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/base/ComplexWidget.java]: Please undo the import changes here (we will avoid * imports)

File: gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+import com.google.gwt.user.client.ui.Widget;
+import gwt.material.design.client.base.ComplexWidget;
+import gwt.material.design.client.base.HasFlexbox;
+import gwt.material.design.client.base.HasTooltip;
+import gwt.material.design.client.base.helper.BrowserPrefixHelper;
+import gwt.material.design.client.constants.*;
+import gwt.material.design.client.ui.MaterialTooltip;
+
+/**
+ * Mixin for the {@link MaterialTooltip} component.
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java]: Need to update comment

**Comment by BenDol** [comment]
Thanks for the contribution so far. I think its important that we think about this implementation being extracted from the ComplexWidget to avoid bloating it too much more. I think Flex layout would suit being its own widget type (though I don't have a lot of experience with it myself).

**Comment by chriswjones** [comment]
Hi @BenDol thanks for the comments. I will fix the imports / other cleanup stuff (IntelliJ gets a little aggressive sometimes).

I am also happy to discuss the best way to implement this.  I have been using flexbox on a few of my projects and have a decent amount of experience / use of it.  There are two main pieces to using flex: the container (flex direction, alignment, justification, etc.) and the children (grow, shrink, basis).  Technically anything can be a container (although usually a div) and anything can be a child.  This is why having the methods on the ComplexWidget is useful. Also the impl has been put into the Flexbox mixin, so it can be reused if necessary.

Reimplementing Display is definitely less than ideal, but I've found it (or something similar) necessary to set display:flex and handle hiding/showing the widget (when setVisible(true) it will not retain the flex display). I am definitely open to suggestions.

File: gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java [hunk 6]
```diff
@@ -0,0 +1,153 @@
...
+import gwt.material.design.client.constants.FlexJustifyContent;
+import gwt.material.design.client.constants.FlexWrap;
+
+/**
+ * Mixin for Flexbox layout
+ *
+ * @author chriswjones
+ */
+public class FlexboxMixin<T extends Widget & HasFlexbox> extends AbstractMixin<T> implements HasFlexbox {
+
```
Comment (BenDol) [inline gwt-material/src/main/java/gwt/material/design/client/base/mixin/FlexboxMixin.java]: Remove extra new line

**Comment by chriswjones** [comment]
removed extra new line

**Comment by gilberto-torrezan** [comment]
Back to the old, ancient Java AWT/Swing framework, this would be a [LayoutManager](https://docs.oracle.com/javase/7/docs/api/java/awt/LayoutManager.html) to be added to any [Container](https://docs.oracle.com/javase/7/docs/api/java/awt/Container.html). The layout manager would deal with the children components, with no modifications needed on any component added to the panel.

Maybe this approach could be used to implement the FlexBox: a container which is responsible for both the parent and the children configuration. Something similar to this approach already exists on GWT, on the [Vertical/Horizontal](http://www.gwtproject.org/javadoc/latest/com/google/gwt/user/client/ui/VerticalPanel.html) panels, which manage a inner table to layout the children components.

(Before any designer comes in kicking the door, breaking the computers and finger-pointing on screams that tables shouldn't be used for layout, I used the example for Vertical/Horizontal panels just to illustrate how a container could interact with the inner elements without interfering at the code of those elements)

**Comment by chriswjones** [comment]
Hi @gilberto-torrezan, thanks for the input. I like the idea of attaching a layout manager, but i don't think this fits the model of how flexbox works.  The flexbox containers set their display type (row or column) and the children handle requesting how much space they want to take up via the grow, shrink and basis properties. This does a better job explaining than i can do:  https://css-tricks.com/snippets/css/a-guide-to-flexbox/

**Comment by gilberto-torrezan** [comment]
Following the approach from VerticalPanels, I imagine something like this:

```java
FlexPanel panel = new FlexPanel(); //the container
panel.setFlexDirection(FlexDirection.ROW); //sets any desired properties on the parent
panel.setFlexWrap(FlexWrap.WRAP);

Widget component = new MaterialPanel(); //the child
panel.add(component); //simple add
panel.setOrder(component, 1); //sets any desired properties on the child, without changing the child implementation
panel.setFlexGrow(component, 1);
//and so on...
```
Utility methods could be added to apply properties to all children, or varargs to add a property to multiple children.

Some black magic would be needed for UiBinder though.

**Comment by chriswjones** [comment]
i like the idea, but the lack of uibinder support would definitely be a bummer. I have used and come to really like having the flex methods (as well as box model methods) on whatever the base control is. This gives a lot of flexibility when designing custom widgets and not needing to add unneeded containers to achieve a certain layout. See below, this is from another project of mine where Input and Div both extend a BasePanel that has flex, box model, focus, etc on it.

``` java
 <a:Div>
        <a:Input ui:field="line1" label="{msg.Line1}"/>
        <a:Input ui:field="line2" label="{msg.Line2}"/>
        <a:Input ui:field="line3" label="{msg.Line3}"/>
        <a:Div flexDirection="ROW">
            <a:Input ui:field="city" label="{msg.City}" flex="ONE" flexShrink="1" marginRight="2px"/>
            <a:Input ui:field="state" label="{msg.State}" width="50px" margin="0 2px"/>
            <a:Input ui:field="postalCode" label="{msg.PostalCode}" width="100px" margin="0 2px"/>
            <a:Input ui:field="country" label="{msg.Country}" flex="ONE" flexShrink="1" marginLeft="2px"/>
        </a:Div>
    </a:Div>
```

I understand the need for the ComplexWidget to remain as small as possible, but these seem like logical things to want to have on all controls. Also I really like your mixin pattern, as that will allow the GWT compiler to optimize each subclass and not include things like the flexbox impl unless they are actually used.

**Comment by BenDol** [comment]
@gilberto-torrezan I was thinking the same thing. But I see the need to update individual components with this strategy also. Will let this sit for a little longer before making any decisions. Thanks @chriswjones

**Comment by chriswjones** [comment]
Glad to help, and thanks for all the input and discussion. I look forward to helping any way I can.

**Comment by kevzlou7979** [comment]
@chriswjones Thank you for your contribution

**Comment by gilberto-torrezan** [comment]
Well, another way of implementing this, following the @chriswjones code and similar approaches on other components of GWT material (such as MaterialTab and MaterialCard), is to create a separate component for the children:

```java
FlexPanel panel = new FlexPanel(); //the container
panel.setFlexDirection(FlexDirection.ROW); //sets any desired properties on the parent
panel.setFlexWrap(FlexWrap.WRAP);

Widget component = new MaterialPanel(); //the component to me added
FlexPanelItem item = new FlexPanelItem(); //the child
panel.add(item); //adds the child in the container
item.add(component); //adds the component to the child
item.setOrder(1); //sets any desired flex properties on the child
item.setFlexGrow(1);
//and so on...
```

When using UiBinder, we would have:

```xml
<m:FlexPanel flexDirection="ROW" flexWrap="WRAP">
    <m:FlexPanelItem order="1" flexGrow="1">
        <m:MaterialPanel />
    </m:FlexPanelItem>
</m:FlexPanel>
```

Of course the names of the classes could be something else (like MaterialFlexBox, MaterialFlexPanel, MaterialAtLastWeDontNeedTablesAnymore, I don't know).

**Comment by chriswjones** [comment]
@gilberto-torrezan i like the idea, but I think it would have two downsides:
- a lot of unnecessary DOM elements created (flex was designed to be attached directly to dom elements
- While the FlexPanelItem you proposed would lay itself out correctly, the children (actual controls inside it) might not as they will still be using box model,etc to handle their layout within. The flexbox model is predicated on the children requesting size (based on their size or flex properties applied)

**Comment by chriswjones** [comment]
also just a heads up I will be creating a different pull request for this later today. I have a few cleanup updates and am getting my fork / branches in order.

**Comment by gilberto-torrezan** [comment]
FlexPanelItem doesn't have to translate to a DOM element. It can be just a marker Widget that sends properties to its children, just like the MaterialTooltip Widget does.

**Comment by chriswjones** [comment]
@gilberto-torrezan you are correct, sorry didn't think of that.  I like the idea, for some reason I really prefer the property route as it seems to mirror the underlying design of flexbox (flexcontainers can often be nested and then be children as well as containers).  

My concern is that by adding the need to wrap elements it makes this verbose / harder to maintain in uibinder and requires the dev to learn our way of flexbox rather than setting properties that mirror the css they are familiar with.

**Comment by chriswjones** [comment]
I know this is still up for debate, just getting my stuff cleaned up. See PR #183
